Configuration file support 🎉

OndraM opened the pull request:
Finally this long-awaited feature is here! Useful for global Steward configuration which doesn't change for different runs.

Usage: place `steward.yml` or `steward.yml.dist` to base directory or use `-c`/`--configuration` option to define custom path to configuration file. Supported options are currently:
 
- `capabilities_resolver` (when `--capability` CLI option is not enough for you, you can implement custom class of `CapabilitiesResolverInterface` which could resolve  capabilities based on any logic you want - see [wiki page](https://github.com/lmc-eu/steward/wiki/Set-custom-capabilities) for more information).
- `tests_dir` (similar to --tests-dir CLI option)
- `logs_dir` (similar to --logs-dir CLI option)
- `fixtures_dir` (similar to --fixtures-dir CLI option)

The `*dir` options could still be defined as a CLI option (and this value will override the value in configuration file).

Some other configuration options will be added in near future (most notably some to allo allow simpler extendability of Steward).

Documentation: https://github.com/lmc-eu/steward/wiki/Configuration-file

OndraM committed:
Improve phpdoc

OndraM committed:
Global settings could be set using configuration file

OndraM committed:
Add option to define custom path to configuration file

OndraM committed:
Trigger new command.pre_initialize event before command initialization is started

OndraM committed:
Update readme

OndraM commented:
New wiki page describing how to use custom capabilities resolver is here: https://github.com/lmc-eu/steward/wiki/Set-custom-capabilities

MortalFlesh commented on src/Console/Configuration/ConfigResolver.php:
```diff
@@ -0,0 +1,113 @@
+        $this->inputDefinition = $inputDefinition;
+    }
+
+    public function resolve(InputInterface $input, array $configFileInputOptions)
```
missing annotation

OndraM replied:
fixed

MortalFlesh commented on src/Console/Configuration/ConfigResolver.php:
```diff
@@ -0,0 +1,113 @@
+                $currentValue = $cliValue;
+            }
+
+            $this->throwExceptionIfDirNotReadable($currentValue, $dirOption);
```
why not to use assertion? or at least name this method `assert` instead of `throwExceptionIf` ?
and btw called method `throwExceptionIfDirNotReadable` should be defined right after first call

OndraM replied:
Good idea to tuse the Assertion library. Fixed. And method renamed (but still kept - the error message assembling is long).

The method is defined ASAP, the previous private method is also called from the parent one (and earlier).

MortalFlesh commented on src/Console/Configuration/OptionsResolverConfigurator.php:
```diff
@@ -0,0 +1,74 @@
+
+        $optionsResolver->setAllowedValues(ConfigOptions::CAPABILITIES_RESOLVER, function ($value) {
+            // Note we throw an exception instead of returning false - to have more understandable exception message
+            if (!class_exists($value)) {
```
`Assertion::classExists` ?
or is it because of specific OptionsResolver exception?

OndraM replied:
Yes, this is just a workaround to overwrite the exception custom message and mimic the behavior of the resolve method - which will throw this exception by itself if this callback returns false. 

:-X

MortalFlesh commented on src/Console/Configuration/OptionsResolverConfigurator.php:
```diff
@@ -0,0 +1,74 @@
+                );
+            }
+
+            if (!is_subclass_of($value, CapabilitiesResolverInterface::class)) {
```
why not `!$value instanceof CapabilitiesResolverInterface::class`?

OndraM replied:
Because the `$value` contains string with FQCN  - not the instance.

jirinovak commented on src/Console/Configuration/ConfigOptions.php:
```diff
@@ -0,0 +1,18 @@
+/**
+ * @codeCoverageIgnore
+ */
+class ConfigOptions
```
final?

jirinovak commented on src/Console/Configuration/ConfigResolver.php:
```diff
@@ -0,0 +1,112 @@
+                $currentValue = $cliValue;
+            } elseif ($configFileValue !== null) {
+                $currentValue = $configFileValue;
+            } else { // In not passed and not in config file use the default CLI value
```
Typo `if`

OndraM committed:
Separate default capabilities resolving from custom ones with CustomCapabilitiesResolverInterface

This should be better than inheriting the default CapabilitiesResolver class (which was also now made final).

OndraM committed:
Remove now unnecessary methods from CapabilitiesResolver

OndraM committed:
Require symfony/process 3.2.0 to make setting empty env variable working properly

vaceletm committed:
Pass custom log dir to PHPUnit process, so that junit files are written there

run allow to specify a custom logs dir but this directory is not taken into
account when running phpunit afterward.

This is useful when you don't have write access to the directory where tests
are defined & run

OndraM committed:
Normalize dir options to never contain trailing slash
